src/pytorch_lightning/plugins/io/torch_plugin.py

```diff
@@ -13,21 +13,52 @@
 # limitations under the License.
 import logging
 import os
+import queue
 from typing import Any, Callable, Dict, Optional
 
 import pytorch_lightning as pl
 from pytorch_lightning.plugins.io.checkpoint_plugin import CheckpointIO
-from pytorch_lightning.utilities.cloud_io import atomic_save, get_filesystem
+from pytorch_lightning.utilities.cloud_io import _atomic_save, atomic_save, get_filesystem
 from pytorch_lightning.utilities.cloud_io import load as pl_load
-from pytorch_lightning.utilities.rank_zero import rank_zero_warn
+from pytorch_lightning.utilities.cloud_io import ThreadQueue
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.rank_zero import rank_zero_info, rank_zero_warn
 from pytorch_lightning.utilities.types import _PATH
 
 log = logging.getLogger(__name__)
 
 
 class TorchCheckpointIO(CheckpointIO):
     """CheckpointIO that utilizes :func:`torch.save` and :func:`torch.load` to save and load checkpoints
-    respectively, common for most use cases."""
+    respectively, common for most use cases.
+
+    Args:
+        save_async: whether to save the checkpoint asynchronously or not.
+        num_threads: Number of threads to use for asynchronous checkpointing.
+    """
+
+    def __init__(self, save_async: bool = False, num_threads: Optional[int] = None):
+        if save_async and not (isinstance(num_threads, int) and (num_threads > 0)):
+            raise MisconfigurationException(
+                f"Asynchronous checkpoint is not possible with `num_threads={num_threads!r}`."
+            )
+
+        if save_async:
+            rank_zero_info(
+                "With asynchronous checkpointing, it is recommended to set"
+                " `ModelCheckpoint(save_top_k=-1)` if your checkpoint interval is low."
+            )
+
+        self.queue = None
+        self.threads = None
```

Comment:
I don't understand why we are using multiple threads. How does this speed up saving? Unless we are saving to a remote location by actually establishing independent connections, we can't scale IO like this just by having more writers. Or how does it work?
I highly recommend only having one thread that pulls from a queue (list) of checkpoints. This is easier to reason about and test as we can guarantee the order in which files are being saved.

Reply:
yep.. will update to single thread